File: packages/sdk-flutter/lib/src/variable_vm.dart
import "models.dart";

class VariableVm {
  dynamic evaluate(CompiledVariable variable, Map<String, dynamic> payload, Map<String, dynamic> variables) {
    final registers = <String, dynamic>{
      "payload": payload,
      "variables": variables,
      "context": variables,
      "source": payload[variable.sourceId] is Map<String, dynamic> ? payload[variable.sourceId] : payload,
    };

    for (final instruction in variable.instructions) {
      final op = instruction.op.toLowerCase();
      final target = instruction.target;
      switch (op) {
        case "literal":
          if (target != null) {
            registers[target] = instruction.value;
          }
          break;
        case "get":
          if (target != null) {
            final source = _resolveRegister(instruction.source, registers) ?? registers["source"];
            registers[target] = _resolvePath(source, instruction.path ?? instruction.key) ?? instruction.defaultValue;
          }
          break;
        case "filter":
          if (target != null) {
            final source = _asList(_resolveRegister(instruction.source, registers));
            registers[target] = source
                .where((item) => _evaluatePredicate(
                      item,
                      instruction.predicatePath ?? instruction.path,
                      instruction.predicateOperator,
                      instruction.predicateValue ?? instruction.value,
                    ))
                .toList();
          }
          break;
        case "find":
          if (target != null) {
            final source = _asList(_resolveRegister(instruction.source, registers));
            registers[target] = source.cast<dynamic?>().firstWhere(
                  (item) => _evaluatePredicate(
                    item,
                    instruction.predicatePath ?? instruction.path,
                    instruction.predicateOperator,
                    instruction.predicateValue ?? instruction.value,
                  ),
                  orElse: () => null,
                );
          }
          break;
        case "sum":
          if (target != null) {
            registers[target] = _extractNumericValues(instruction, registers).fold<double>(0, (sum, value) => sum + value);
          }
          break;
        case "count":
          if (target != null) {
            final value = _resolveRegister(instruction.source, registers);
            registers[target] = switch (value) {
              List<dynamic> => value.length,
              Map<String, dynamic> => value.length,
              null => 0,
              _ => 1,
            };
          }
          break;
        case "count_where":
          if (target != null) {
            final source = _asList(_resolveRegister(instruction.source, registers));
            registers[target] = source
                .where((item) => _evaluatePredicate(
                      item,
                      instruction.predicatePath ?? instruction.path,
                      instruction.predicateOperator,
                      instruction.predicateValue ?? instruction.value,
                    ))
                .length;
          }
          break;
        case "max":
          if (target != null) {
            final values = _extractNumericValues(instruction, registers);
            registers[target] = values.isEmpty ? 0.0 : values.reduce((left, right) => left > right ? left : right);
          }
          break;
        case "min":
          if (target != null) {
            final values = _extractNumericValues(instruction, registers);
            registers[target] = values.isEmpty ? 0.0 : values.reduce((left, right) => left < right ? left : right);
          }
          break;
        case "divide":
        case "multiply":
        case "add":
        case "subtract":
          if (target != null) {
            final left = _numeric(_resolveOperand(instruction.left, registers, instruction.values.isNotEmpty ? instruction.values.first : null));
            final right = _numeric(
              _resolveOperand(
                instruction.right,
                registers,
                instruction.values.length > 1 ? instruction.values[1] : instruction.value ?? instruction.defaultValue,
              ),
            );
            registers[target] = switch (op) {
              "divide" => right == 0 ? 0.0 : left / right,
              "multiply" => left * right,
              "subtract" => left - right,
              _ => left + right,
            };
          }
          break;
        case "cast":
          if (target != null) {
            registers[target] = _castValue(_resolveRegister(instruction.source, registers), instruction.type);
          }
          break;
        case "compare":
          if (target != null) {
            registers[target] = _compare(
              _resolveOperand(instruction.left, registers, null),
              instruction.predicateOperator ?? instruction.type ?? "==",
              _resolveOperand(instruction.right, registers, instruction.value),
            );
          }
          break;
        case "if":
          if (target != null) {
            final condition = _resolveRegister(instruction.source, registers);
            final branch = condition is bool ? condition : _numeric(condition) != 0;
            registers[target] = branch
                ? _resolveOperand(instruction.thenSource, registers, instruction.thenValue)
                : _resolveOperand(instruction.elseSource, registers, instruction.elseValue);
          }
          break;
        case "return":
          return _resolveOperand(instruction.source ?? target, registers, instruction.value);
      }
    }
    return null;
  }

  dynamic _resolveRegister(String? name, Map<String, dynamic> registers) {
    if (name == null || name.isEmpty) {
      return null;
    }
    if (registers.containsKey(name)) {
      return registers[name];
    }
    return _resolvePath(registers["source"], name) ?? _resolvePath(registers["payload"], name) ?? _resolvePath(registers["variables"], name);
  }

  dynamic _resolveOperand(String? name, Map<String, dynamic> registers, dynamic fallback) => _resolveRegister(name, registers) ?? fallback;

  List<double> _extractNumericValues(Instruction instruction, Map<String, dynamic> registers) {
    final directSources = instruction.sources.isNotEmpty
        ? instruction.sources
        : <String>[
            if (instruction.source != null && instruction.source!.isNotEmpty) instruction.source!,
            ...instruction.args,
          ];
    if (directSources.isNotEmpty) {
      return directSources.map((source) => _numeric(_resolveRegister(source, registers))).toList();
    }
    final list = _asList(_resolveRegister(instruction.source, registers));
    return list
        .map((item) => instruction.path == null || instruction.path!.isEmpty ? item : _resolvePath(item, instruction.path))
        .map(_numeric)
        .toList();
  }

  List<dynamic> _asList(dynamic value) {
    if (value == null) {
      return const <dynamic>[];
    }
    if (value is List<dynamic>) {
      return value;
    }
    return <dynamic>[value];
  }

  dynamic _resolvePath(dynamic root, String? rawPath) {
    if (root == null || rawPath == null || rawPath.isEmpty) {
      return root;
    }
    final path = rawPath.startsWith(r"$.") ? rawPath.substring(2) : rawPath;
    if (path.isEmpty) {
      return root;
    }
    dynamic current = root;
    for (final segment in path.split(".")) {
      if (current == null) {
        return null;
      }
      final match = RegExp(r"([A-Za-z0-9_\-]+)(\[(\-?\d+)\])?").firstMatch(segment);
      if (match == null) {
        return null;
      }
      final key = match.group(1)!;
      current = current is Map<String, dynamic> ? current[key] : null;
      final indexGroup = match.group(3);
      if (indexGroup != null) {
        final list = current is List<dynamic> ? current : null;
        if (list == null) {
          return null;
        }
        final requested = int.parse(indexGroup);
        final resolvedIndex = requested < 0 ? list.length + requested : requested;
        current = resolvedIndex >= 0 && resolvedIndex < list.length ? list[resolvedIndex] : null;
      }
    }
    return current;
  }

  bool _evaluatePredicate(dynamic item, String? path, String? operator, dynamic expected) {
    final actual = path == null || path.isEmpty ? item : _resolvePath(item, path);
    return _compare(actual, operator ?? "==", expected);
  }

  bool _compare(dynamic actual, String operator, dynamic expected) {
    switch (operator) {
      case ">":
        return _numeric(actual) > _numeric(expected);
      case ">=":
        return _numeric(actual) >= _numeric(expected);
      case "<":
        return _numeric(actual) < _numeric(expected);
      case "<=":
        return _numeric(actual) <= _numeric(expected);
      case "!=":
        return _normalize(actual) != _normalize(expected);
      case "in":
        return _asList(expected).map(_normalize).contains(_normalize(actual));
      case "not_in":
        return !_asList(expected).map(_normalize).contains(_normalize(actual));
      default:
        return _normalize(actual) == _normalize(expected);
    }
  }

  dynamic _castValue(dynamic value, String? type) {
    switch ((type ?? "string").toLowerCase()) {
      case "int":
      case "integer":
        return _numeric(value).toInt();
      case "float":
      case "double":
      case "number":
        return _numeric(value);
      case "bool":
      case "boolean":
        if (value is bool) {
          return value;
        }
        if (value is num) {
          return value != 0;
        }
        if (value is String) {
          return value.toLowerCase() == "true" || value == "1";
        }
        return value != null;
      case "list":
        return _asList(value);
      default:
        return value?.toString();
    }
  }

  dynamic _normalize(dynamic value) {
    if (value is num) {
      return value.toDouble();
    }
    if (value is String) {
      return value.trim();
    }
    return value;
  }

  double _numeric(dynamic value) {
    if (value is num) {
      return value.toDouble();
    }
    if (value is bool) {
      return value ? 1.0 : 0.0;
    }
    if (value is String) {
      return double.tryParse(value) ?? 0.0;
    }
    return 0.0;
  }
}

File: packages/sdk-android/rulemind-core/src/main/kotlin/com/rulemind/core/VariableVM.kt
package com.rulemind.core

import com.rulemind.core.models.CompiledVariable
import com.rulemind.core.models.Instruction
import java.util.Locale

class VariableVM {
    fun evaluate(variable: CompiledVariable, payload: Map<String, Any?>, variables: Map<String, Any?>): Any? {
        val registers = mutableMapOf<String, Any?>(
            "payload" to payload,
            "variables" to variables,
            "context" to variables,
        )
        val sourcePayload = (payload[variable.sourceId] as? Map<String, Any?>) ?: payload
        registers["source"] = sourcePayload

        for (instruction in variable.instructions) {
            val target = instruction.target
            when (instruction.op.lowercase(Locale.ROOT)) {
                "literal" -> if (target != null) {
                    registers[target] = instruction.value
                }

                "get" -> if (target != null) {
                    val source = resolveRegister(instruction.source, registers) ?: sourcePayload
                    registers[target] = resolvePath(source, instruction.path ?: instruction.key) ?: instruction.defaultValue
                }

                "filter" -> if (target != null) {
                    val source = asList(resolveRegister(instruction.source, registers))
                    registers[target] = source.filter { item ->
                        evaluatePredicate(
                            item,
                            instruction.predicatePath ?: instruction.path,
                            instruction.predicateOperator,
                            instruction.predicateValue ?: instruction.value,
                        )
                    }
                }

                "find" -> if (target != null) {
                    val source = asList(resolveRegister(instruction.source, registers))
                    registers[target] = source.firstOrNull { item ->
                        evaluatePredicate(
                            item,
                            instruction.predicatePath ?: instruction.path,
                            instruction.predicateOperator,
                            instruction.predicateValue ?: instruction.value,
                        )
                    }
                }

                "sum" -> if (target != null) {
                    registers[target] = extractNumericValues(instruction, registers).sum()
                }

                "count" -> if (target != null) {
                    val value = resolveRegister(instruction.source, registers)
                    registers[target] = when (value) {
                        is Collection<*> -> value.size
                        is Map<*, *> -> value.size
                        null -> 0
                        else -> 1
                    }
                }

                "count_where" -> if (target != null) {
                    val source = asList(resolveRegister(instruction.source, registers))
                    registers[target] = source.count { item ->
                        evaluatePredicate(
                            item,
                            instruction.predicatePath ?: instruction.path,
                            instruction.predicateOperator,
                            instruction.predicateValue ?: instruction.value,
                        )
                    }
                }

                "max" -> if (target != null) {
                    registers[target] = extractNumericValues(instruction, registers).maxOrNull() ?: 0.0
                }

                "min" -> if (target != null) {
                    registers[target] = extractNumericValues(instruction, registers).minOrNull() ?: 0.0
                }

                "divide", "multiply", "add", "subtract" -> if (target != null) {
                    val left = numeric(resolveOperand(instruction.left, registers, instruction.values.getOrNull(0)))
                    val right = numeric(
                        resolveOperand(
                            instruction.right,
                            registers,
                            instruction.values.getOrNull(1) ?: instruction.value ?: instruction.defaultValue,
                        )
                    )
                    registers[target] = when (instruction.op.lowercase(Locale.ROOT)) {
                        "divide" -> if (right == 0.0) 0.0 else left / right
                        "multiply" -> left * right
                        "subtract" -> left - right
                        else -> left + right
                    }
                }

                "cast" -> if (target != null) {
                    registers[target] = castValue(resolveRegister(instruction.source, registers), instruction.type)
                }

                "compare" -> if (target != null) {
                    registers[target] = compare(
                        resolveOperand(instruction.left, registers, null),
                        instruction.predicateOperator ?: instruction.type ?: "==",
                        resolveOperand(instruction.right, registers, instruction.value),
                    )
                }

                "if" -> if (target != null) {
                    val conditionValue = resolveRegister(instruction.source, registers)
                    val branch = when (conditionValue) {
                        is Boolean -> conditionValue
                        is Number -> conditionValue.toDouble() != 0.0
                        null -> false
                        else -> true
                    }
                    registers[target] = if (branch) {
                        resolveOperand(instruction.thenSource, registers, instruction.thenValue)
                    } else {
                        resolveOperand(instruction.elseSource, registers, instruction.elseValue)
                    }
                }

                "return" -> return resolveOperand(instruction.source ?: target, registers, instruction.value)
            }
        }
        return null
    }

    private fun resolveRegister(name: String?, registers: Map<String, Any?>): Any? {
        if (name.isNullOrBlank()) {
            return null
        }
        if (registers.containsKey(name)) {
            return registers[name]
        }
        return resolvePath(registers["source"], name)
            ?: resolvePath(registers["payload"], name)
            ?: resolvePath(registers["variables"], name)
    }

    private fun resolveOperand(name: String?, registers: Map<String, Any?>, fallback: Any?): Any? {
        return resolveRegister(name, registers) ?: fallback
    }

    private fun extractNumericValues(instruction: Instruction, registers: Map<String, Any?>): List<Double> {
        val directSources = instruction.sources.ifEmpty {
            buildList {
                if (!instruction.source.isNullOrBlank()) add(instruction.source)
                addAll(instruction.args)
            }
        }
        if (directSources.isNotEmpty()) {
            return directSources.map { numeric(resolveRegister(it, registers)) }
        }

        val base = resolveRegister(instruction.source, registers)
        val list = asList(base)
        if (list.isEmpty()) {
            return emptyList()
        }
        return list.map { item ->
            numeric(
                if (instruction.path.isNullOrBlank()) {
                    item
                } else {
                    resolvePath(item, instruction.path)
                }
            )
        }
    }

    private fun asList(value: Any?): List<Any?> = when (value) {
        is List<*> -> value
        is Array<*> -> value.toList()
        null -> emptyList()
        else -> listOf(value)
    }

    private fun resolvePath(root: Any?, rawPath: String?): Any? {
        if (root == null || rawPath.isNullOrBlank()) {
            return root
        }
        val path = rawPath.removePrefix("$.")
        if (path.isBlank()) {
            return root
        }
        var current: Any? = root
        val segments = path.split(".")
        for (segment in segments) {
            if (current == null) {
                return null
            }
            val match = Regex("([A-Za-z0-9_\\-]+)(\\[(\\-?\\d+)])?").matchEntire(segment) ?: return null
            val key = match.groupValues[1]
            current = when (current) {
                is Map<*, *> -> current[key]
                else -> null
            }
            val indexGroup = match.groupValues.getOrNull(3).orEmpty()
            if (indexGroup.isNotBlank()) {
                val list = current as? List<*> ?: return null
                val requested = indexGroup.toInt()
                val index = if (requested < 0) list.size + requested else requested
                current = list.getOrNull(index)
            }
        }
        return current
    }

    private fun evaluatePredicate(item: Any?, path: String?, operator: String?, expected: Any?): Boolean {
        val actual = if (path.isNullOrBlank()) item else resolvePath(item, path)
        return compare(actual, operator ?: "==", expected)
    }

    private fun compare(actual: Any?, operator: String, expected: Any?): Boolean {
        val actualNumber = actual as? Number
        val expectedNumber = expected as? Number
        return when (operator) {
            ">" -> numeric(actualNumber) > numeric(expectedNumber)
            ">=" -> numeric(actualNumber) >= numeric(expectedNumber)
            "<" -> numeric(actualNumber) < numeric(expectedNumber)
            "<=" -> numeric(actualNumber) <= numeric(expectedNumber)
            "!=" -> normalize(actual) != normalize(expected)
            "in" -> asList(expected).map(::normalize).contains(normalize(actual))
            "not_in" -> !asList(expected).map(::normalize).contains(normalize(actual))
            else -> normalize(actual) == normalize(expected)
        }
    }

    private fun castValue(value: Any?, type: String?): Any? = when ((type ?: "string").lowercase(Locale.ROOT)) {
        "int", "integer" -> numeric(value).toInt()
        "float", "double", "number" -> numeric(value)
        "bool", "boolean" -> when (value) {
            is Boolean -> value
            is Number -> value.toDouble() != 0.0
            is String -> value.equals("true", ignoreCase = true) || value == "1"
            else -> value != null
        }
        "list" -> asList(value)
        else -> value?.toString()
    }

    private fun normalize(value: Any?): Any? = when (value) {
        is Number -> value.toDouble()
        is String -> value.trim()
        else -> value
    }

    private fun numeric(value: Any?): Double = when (value) {
        is Number -> value.toDouble()
        is String -> value.toDoubleOrNull() ?: 0.0
        is Boolean -> if (value) 1.0 else 0.0
        else -> 0.0
    }
}
